packages/npm-resolver/src/pickPackageFromMeta.ts

```diff
@@ -76,6 +77,17 @@ function pickVersionByVersionRange (
         }
         break
       }
+      // This is a special type of "preference". `versionPref` is the version
+      // range written by yarn into the lockfile. `originalPref` is what pnpm
+      // found in the requesting package.json. If they match, we should exit
+      // early and use the version resolved in the lockfile.
+      case 'lockfile': {
```

Comment:
This is the meat of my change. I added a new `lockfile` version selector. See the comment for more details.